### PR TITLE
`ultralytics 8.3.189` 3x faster inplace `model.fuse()`

### PR DESCRIPTION
**Pull request by glenn-jocher**
## 🛠️ PR Summary

<sub>Made with ❤️ by [Ultralytics Actions](https://github.com/ultralytics/actions)<sub>

### 🌟 Summary
In-place fusion of Conv/BatchNorm and Deconv/BatchNorm layers for faster, lighter inference, plus a version bump to 8.3.189. 🚀

### 📊 Key Changes
- Fuse functions updated:
  - `fuse_conv_and_bn(conv, bn)` now fuses weights and bias directly into the existing `nn.Conv2d` layer (in-place) and returns it with gradients disabled.
  - `fuse_deconv_and_bn(deconv, bn)` does the same for `nn.ConvTranspose2d`.
- Bias handling improved:
  - If the original layer had no bias, the function now registers a fused bias parameter correctly.
  - Corrected bias shape handling for transposed conv fusion using `deconv.out_channels`.
- Added clear docstrings with Args/Returns and simple usage examples.
- Version bump: `8.3.189`.

### 🎯 Purpose & Impact
- Performance boost: Reduces memory allocations and speeds up inference by avoiding creation of new layer objects. ⚡
- Safer and cleaner: Properly handles bias registration and shapes, reducing edge-case bugs (especially for deconvs). 🧰
- API behavior: Functions still return a fused layer, but now it’s the same module updated in-place. This may affect code that relied on receiving a new module instance.
- Intended for inference: Returned layers have gradients disabled, aligning with typical post-training fusion workflows. ✅

Example:
```python
import torch.nn as nn
from ultralytics.utils.torch_utils import fuse_conv_and_bn, fuse_deconv_and_bn

conv, bn = nn.Conv2d(3, 16, 3), nn.BatchNorm2d(16)
conv = fuse_conv_and_bn(conv, bn)

deconv, bn2 = nn.ConvTranspose2d(16, 3, 3), nn.BatchNorm2d(3)
deconv = fuse_deconv_and_bn(deconv, bn2)
```

### CODE REVIEW

**Comment by UltralyticsAssistant** [comment]
👋 Hello @glenn-jocher, thank you for submitting an `ultralytics/ultralytics` 🚀 PR! This is an automated response to help streamline reviews. An Ultralytics engineer will also assist you here shortly. To ensure a seamless integration of your work on passing the device directly in fuse ops, please review the following checklist:

- ✅ **Define a Purpose**: Clearly explain the purpose of your fix or feature in your PR description, and link to any [relevant issues](https://github.com/ultralytics/ultralytics/issues). Ensure your commit messages are clear, concise, and adhere to the project's conventions.
- ✅ **Synchronize with Source**: Confirm your PR is synchronized with the `ultralytics/ultralytics` `main` branch. If it's behind, update it by clicking the 'Update branch' button or by running `git pull` and `git merge main` locally.
- ✅ **Ensure CI Checks Pass**: Verify all Ultralytics [Continuous Integration (CI)](https://docs.ultralytics.com/help/CI/) checks are passing. If any checks fail, please address the issues.
- ✅ **Update Documentation**: Update the relevant [documentation](https://docs.ultralytics.com/) for any new or modified features.
- ✅ **Add Tests**: If applicable, include or update tests to cover your changes, and confirm that all tests are passing.
- ✅ **Sign the CLA**: Please ensure you have signed our [Contributor License Agreement](https://docs.ultralytics.com/help/CLA/) if this is your first Ultralytics PR by writing "I have read the CLA Document and I sign the CLA" in a new message.
- ✅ **Minimize Changes**: Limit your changes to the **minimum** necessary for your bug fix or feature addition. _"It is not daily increase but daily decrease, hack away the unessential. The closer to the source, the less wastage there is."_  — Bruce Lee

Notes specific to this PR:
- Thanks for optimizing fusion to instantiate Conv/ConvTranspose on the target device directly 🎯 Please ensure compatibility across supported PyTorch versions that accept `device=...` in constructors.
- Consider adding or updating unit tests that validate fused layer creation on CPU and CUDA to avoid regressions and ensure parity with prior behavior (bias handling, gradients disabled).
- Double-check that quantization/export paths and AMP/backends remain unaffected by these changes.

For more guidance, please refer to our [Contributing Guide](https://docs.ultralytics.com/help/contributing/). Don’t hesitate to leave a comment if you have any questions. Thank you for contributing to Ultralytics! ⚡️🚀

**Comment by glenn-jocher** [comment]
Well this sucks. I think I might have tried this a couple years ago haha.

**Comment by glenn-jocher** [comment]
@ultralytics/yolo **fuse ops profile about 3x faster** here! Tested with YOLO11m on `yolo predict`: 60ms to 20ms on M4 Pro.

### Main
```
Function                                                     Cum Time Self Time Calls  Avg     
-----------------------------------------------------------------------------------------------
ultralytics.nn.tasks:fuse:224                                0.062    0.001    1      0.06153 
ultralytics.utils.torch_utils:fuse_conv_and_bn:252           0.057    0.003    106    0.00053 
```

### This PR
```
Function                                                     Cum Time Self Time Calls  Avg     
-----------------------------------------------------------------------------------------------
ultralytics.nn.tasks:fuse:224                                0.023    0.001    1      0.02343 
ultralytics.utils.torch_utils:fuse_conv_and_bn:252           0.020    0.002    106    0.00019 
```

**Comment by glenn-jocher** [comment]
Even works on torch 1.8. Super ironic that I only did this because my original attempt failed on 1.8, otherwise I would have just merged that for negligible speedup. Misfortune converts to fortune.

**Comment by glenn-jocher** [comment]
## ⚡ Actions Trigger

<sub>Made with ❤️ by [Ultralytics Actions](https://www.ultralytics.com/actions)<sub>

GitHub Actions below triggered via workflow dispatch on this PR branch `glenn-jocher-patch-1` at 2025-08-28 14:51:36 UTC with `@ultralytics/run-ci` command:

* ✅ [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/runs/17299559819): `ci.yml` run 10902
* ✅ [Publish Docker Images](https://github.com/ultralytics/ultralytics/actions/runs/17299559960): `docker.yml` run 1003

**Comment by Laughing-q** [comment]
@glenn-jocher awesome work! oh now we just modify the weights in-place, instead of creating a new Conv and fed it with the copy of weights and bias.
haha I think we have this for years when you firstly created it years ago. That's why I always want to review the whole codebase.

**Comment by glenn-jocher** [comment]
@Laughing-q yeah it's so much better now just updating the instance instead of creating many new instances of the convolution.

I bet there are plenty of other areas we could improve also, but there's too many places, so we need to focus on the hotspots that appear from high-level profiling like what @onuralpszr was showing and what my example code showed.

**Comment by Laughing-q** [comment]
@glenn-jocher wait, this just came to my mind, what if...what if the fuse process running on gpu is less efficient than on cpu, because of the copy operation?😂
can you do a test for it as well? I just shutdown my pc.

**Comment by glenn-jocher** [comment]
@Laughing-q yeah I can run some profiling in Colab

**Comment by glenn-jocher** [comment]
Ok GPU profiling in Colab with `yolo predict` timing fuse ops:

### Main
```
0.432s
0.131s
0.134s
0.134s
```

### This PR
```
0.340
0.085
0.086
0.086
```

Don't see the same speedup as on my mac on CPU, but still improves, maybe 1/3 faster.

**Comment by Laughing-q** [comment]
@glenn-jocher oh I meant with this PR eliminating the copy operation, probably we don't need the fuse before moving to certain device PR? https://github.com/ultralytics/ultralytics/pull/21853
no need to fuse on cpu before moving to gpu/mps to speed up? or maybe the original version with this PR might even be faster😂?

**Comment by glenn-jocher** [comment]
@Laughing-q not sure. I've created a new revert PR targeted here to test:
https://github.com/ultralytics/ultralytics/pull/21859

**Comment by glenn-jocher** [comment]
@Laughing-q fuse times in https://github.com/ultralytics/ultralytics/pull/21859 are horrible. The user was right, we should have listened to him earlier.

```
1.440
0.452
0.439
0.433
```

**Comment by UltralyticsAssistant** [comment]
Absolutely thrilled to see this merged! Thank you @glenn-jocher and kudos to @Laughing-q for the sharp contributions in PR [In-place Conv/Deconv + BatchNorm fusion for faster inference](https://github.com/ultralytics/ultralytics/pull/21844). 

“Perfection is achieved not when there is nothing more to add, but when there is nothing left to take away.” — Antoine de Saint‑Exupéry. This resonates perfectly here: in-place fusion, leaner memory use, and safer bias handling—all aligned for faster, cleaner inference in 8.3.189. 

Appreciate the clear docstrings and thoughtful API notes. Tiny wins that compound at scale. Thanks for pushing the edge forward!

Quick tip for users:
- Fuse in-place for inference: `conv = fuse_conv_and_bn(conv, bn)` and `deconv = fuse_deconv_and_bn(deconv, bn)` (returns the same module with grads disabled).
